jupyter-messages.el

```diff
@@ -152,38 +152,34 @@ If PART is a list whose first element is the symbol,
 is non-nil.  If it is nil, then set the list's second element to
 the result of calling `jupyter--encode' on the third element and
 return the result."
-  (if (eq (car-safe part) 'message-part)
-      (cl-destructuring-bind (_ encoded-rep decoded-rep) part
-        (or encoded-rep (setf (nth 1 part) (jupyter--encode decoded-rep))))
-    (cl-letf (((symbol-function 'json-encode)
-               (lambda (object)
-                 (cond ((memq object (list t json-null json-false))
-                        (json-encode-keyword object))
-                       ((stringp object)      (json-encode-string object))
-                       ((keywordp object)
-                        ;; Handle `jupyter-message-type'
-                        (let ((msg-type (plist-get jupyter-message-types object)))
-                          (json-encode-string
-                           (or msg-type (substring (symbol-name object) 1)))))
-                       ((symbolp object)      (json-encode-string (symbol-name object)))
-                       ((numberp object)      (json-encode-number object))
-                       ((arrayp object)       (json-encode-array object))
-                       ((hash-table-p object) (json-encode-hash-table object))
-                       ((listp object)
-                        (cond
-                         ((eq (car object) 'message-part)
-                          (jupyter--encode object))
-                         ;; Turn time objects into ISO 8601 time strings
-                         ((and (= (length object) 4)
-                               (cl-every #'integerp object))
-                          (jupyter-encode-time object))
-                         (t (json-encode-list object))))
-                       (t                     (signal 'json-error (list object)))))))
-      (encode-coding-string
-       (cond
-        ((stringp part) part)
-        (t (json-encode part)))
-       'utf-8 t))))
+  (let ((json-recursive-encoder-sym (if (fboundp 'json--print) 'json--print 'json-encode)))
```

Comment:
Can you move the `if` call outside of the function and into a `defvar`?  I think that would be better than checking for `json--print` every time a message gets encoded.

Reply:
I could, but I think the "right" choice in terms of behaviour is checking every time since in principle the user could load the new or old JSON APIs during an Emacs session. Although this is an (extreme) edge case, since the check is inexpensive (especially compared to the rest of the function; note this check is only done once when the JSON encoder is called, not recursively) I have a weak preference to keep it this way.

Let me know if you still want the check moved to a `defvar` in light of this, happy to do so then.

Reply:
Your right, it only happens once per message encoding so it's not a big deal.